

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Changelog &mdash; msprime 0.1.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/jupyter-sphinx.css" type="text/css" />
  <link rel="stylesheet" href="_static/thebelab.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="PORTING IN PROGRESS: Tutorial" href="tutorial.html" />
    <link rel="prev" title="Citing msprime" href="CITATION.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> msprime
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="ancestry.html">Ancestry</a></li>
<li class="toctree-l1"><a class="reference internal" href="mutations.html">Mutations</a></li>
<li class="toctree-l1"><a class="reference internal" href="demography.html">Demography</a></li>
<li class="toctree-l1"><a class="reference internal" href="likelihoods.html">Likelihoods</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickref.html">API quick reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="utilities.html">Utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Developer documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing msprime</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Changelog</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#upcoming-x-x-xxxx-xx-xx">[UPCOMING.X.X] - XXXX-XX-XX</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id1">[0.7.4] - 2019-12-05</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id2">[0.7.3] - 2019-08-03</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id3">[0.7.2] - 2019-07-30</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id4">[0.7.1] - 2019-06-08</a></li>
<li class="toctree-l2"><a class="reference internal" href="#b1-2019-05-31">[0.7.1b1] - 2019-05-31</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id5">[0.7.0] - 2019-02-19</a></li>
<li class="toctree-l2"><a class="reference internal" href="#a1-2019-01-14">[0.7.0a1] - 2019-01-14</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id6">[0.6.2] - 2018-12-04</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id7">[0.6.1] - 2018-08-25</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id8">[0.6.0] - 2018-06-20</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id9">[0.5.0] - 2018-02-26</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id10">[0.4.0] - 2016-10-16</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id11">[0.3.2] - 2016-07-21</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id12">[0.3.1] - 2016-06-24</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id13">[0.3.0] - 2016-05-31</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id14">[0.2.0] - 2016-05-05</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id15">[0.1.10] - 2016-04-21</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id16">[0.1.9] - 2016-04-01</a></li>
<li class="toctree-l2"><a class="reference internal" href="#id17">[0.1.8] - 2016-02-17</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#changes">Changes</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">PORTING IN PROGRESS: Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">PORTING IN PROGRESS: API</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">msprime</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Changelog</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/changelog.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="changelog">
<span id="sec-changelog"></span><h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline">¶</a></h1>
<div class="section" id="upcoming-x-x-xxxx-xx-xx">
<h2>[UPCOMING.X.X] - XXXX-XX-XX<a class="headerlink" href="#upcoming-x-x-xxxx-xx-xx" title="Permalink to this headline">¶</a></h2>
<p><strong>New features</strong>:</p>
<ul class="simple">
<li><p>Add functions to parse species trees and set up simulation models according
to the species tree. (<a class="reference external" href="https://github.com/mmatschiner">&#64;mmatschiner</a> <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/893">#893</a> <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/929">#929</a> <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/931">#931</a>)</p></li>
<li><p>Complete provenance recording of all arguments to simulate and mutate.
Adds argument record_provenance to simulate, which allows recording of
provenances to be disabled, for example when they are large.
(<a class="reference external" href="https://github.com/benjeffery">&#64;benjeffery</a> <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/914">#914</a>).</p></li>
<li><p>Add replicate_index to simulate, allowing output of a single tree sequence
from a set of replicates. (<a class="reference external" href="https://github.com/benjeffery">&#64;benjeffery</a> <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/914">#914</a>).</p></li>
<li><p>Much better simulation performance for models with large numbers
of populations (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/1069">#1069</a>).</p></li>
<li><p>Details of the simulation are written to the DEBUG log periodically.
This can help debug long-running simulations. (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>,
<a class="reference external" href="https://github.com/tskit-dev/msprime/pull/1080">#1080</a>).</p></li>
<li><p>Add a <code class="docutils literal notranslate"><span class="pre">discrete_coordinates</span></code> argument to <cite>simulate</cite>, which ensures that
all recombination breakpoints and mutational sites are at integer positions
(<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/880">#880</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/1115">#1115</a>).</p></li>
</ul>
<p><strong>Breaking changes</strong>:</p>
<ul class="simple">
<li><p>The class form for specifying models (e.g., <code class="docutils literal notranslate"><span class="pre">msprime.StandardCoalescent()</span></code>)
no longer take a <code class="docutils literal notranslate"><span class="pre">reference_size</span></code> argument.
(TODO add paper trail for this)</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">simulate</span></code> function only takes one positional argument, and all other
arguments are keyword-only.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">msp</span></code> CLI has been stripped of all sub-commands except for
<code class="docutils literal notranslate"><span class="pre">simulate</span></code> and <code class="docutils literal notranslate"><span class="pre">mutate</span></code>. These sub-commands are provided by the <code class="docutils literal notranslate"><span class="pre">tskit</span></code>
CLI or the <code class="docutils literal notranslate"><span class="pre">TreeSequence</span></code> API in <code class="docutils literal notranslate"><span class="pre">tskit</span></code>.</p></li>
<li><p>Population metadata is now automatically decoded in tree sequences
produced by msprime. This means that code doing things like
<code class="docutils literal notranslate"><span class="pre">json.loads(population.metadata)</span></code> will now break. To recover the
original metadata passed to PopulationConfiguration objects, we
can write <code class="docutils literal notranslate"><span class="pre">population.metadata[&quot;user&quot;]</span></code>.</p></li>
</ul>
<p><strong>Deprecations</strong>:</p>
<ul class="simple">
<li><p>Deprecate module attributes that were moved to tskit.
(<a class="reference external" href="https://github.com/benjeffery">&#64;benjeffery</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/991">#991</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/1158">#1158</a>)</p></li>
</ul>
</div>
<div class="section" id="id1">
<h2>[0.7.4] - 2019-12-05<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h2>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Fix error in mspms output of tree spans. In previous versions, the length of
genome spanned by trees in the newick output was incorrect in certain situations
(specifically, when “invisible” recombinations are present so that two or more
identical trees are printed out). Thanks to <a class="reference external" href="https://github.com/fbaumdicker">&#64;fbaumdicker</a> for spotting
the problem. (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/837">#837</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/836">#836</a>)</p></li>
<li><p>Fix assertion tripped when we have very low recombination rates in the DTWF
model. Thanks to <a class="reference external" href="https://github.com/terhorst">&#64;terhorst</a> for the bug report.
(<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/833">#833</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/831">#831</a>).</p></li>
<li><p>Fix bug in memory allocation when simulating mutations on a tree sequence
that already contains many mutations. Thanks to <a class="reference external" href="https://github.com/santaci">&#64;santaci</a> for the
bug report. (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>, <a class="reference external" href="https://github.com/petrelharp">&#64;petrelharp</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/838">#838</a>,
<a class="reference external" href="https://github.com/tskit-dev/msprime/issues/806">#806</a>)</p></li>
</ul>
<p><strong>New features</strong>:</p>
<ul class="simple">
<li><p>Add the new Census event, which allows us to place nodes on all extant
branches at a given time (<a class="reference external" href="https://github.com/gtsambos">&#64;gtsambos</a> <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/799">#799</a>).</p></li>
<li><p>Improved error reporting for input parameters, in particular
demographic events (<a class="reference external" href="https://github.com/tskit-dev/msprime/pull/829">#829</a>).</p></li>
</ul>
<p><strong>Documentation</strong>:</p>
<ul class="simple">
<li><p>Improved container documentation (<a class="reference external" href="https://github.com/agladstein">&#64;agladstein</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/822">#822</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/809">#809</a>).</p></li>
<li><p>Improved developer docs for macs (<a class="reference external" href="https://github.com/gtsambos">&#64;gtsambos</a>, <a class="reference external" href="https://github.com/molpopgen">&#64;molpopgen</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/805">#805</a>).</p></li>
<li><p>Clarify meaning of migration matrix (<a class="reference external" href="https://github.com/petrelharp">&#64;petrelharp</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/830">#830</a>).</p></li>
</ul>
</div>
<div class="section" id="id2">
<h2>[0.7.3] - 2019-08-03<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Support for SMC models coupled with the record_full_arg feature was
erroneously removed in a previous version (<a class="reference external" href="https://github.com/tskit-dev/msprime/issues/795">#795</a>). The feature
has been resinstated (<a class="reference external" href="https://github.com/tskit-dev/msprime/pull/796">#796</a>).</p></li>
</ul>
</div>
<div class="section" id="id3">
<h2>[0.7.2] - 2019-07-30<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h2>
<p><strong>Breaking changes</strong></p>
<ul class="simple">
<li><p>The random trajectory has been changed slightly to improve handling
of ancient sampling events (<a class="reference external" href="https://github.com/tskit-dev/msprime/pull/782">#782</a>). Thus, simulations for a given
random seed will not be identical to previous versions, if ancient
samples are used.</p></li>
</ul>
<p><strong>New features</strong></p>
<ul class="simple">
<li><p>Automated Docker builds (<a class="reference external" href="https://github.com/agladstein">&#64;agladstein</a>; <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/661">#661</a>)</p></li>
<li><p>Add mean coalescence time to DemographyDebugger (<a class="reference external" href="https://github.com/petrelharp">&#64;petrelharp</a>; <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/779">#779</a>).</p></li>
<li><p>Improve MassMigration descriptions in DemographyDebugger
(<a class="reference external" href="https://github.com/marianne-aspbury">&#64;marianne-aspbury</a>; <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/791">#791</a>).</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>In very, very, very rare cases it was possible to generate a
zero waiting time until the next coalescent event, leading to
zero branch lengths in the output tree sequence and an error
being raised (<a class="reference external" href="https://github.com/molpopgen">&#64;molpopgen</a>, <a class="reference external" href="https://github.com/DL42">&#64;DL42</a>, <a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>;
<a class="reference external" href="https://github.com/tskit-dev/msprime/issues/783">#783</a>, <a class="reference external" href="https://github.com/tskit-dev/msprime/pull/785">#785</a>).</p></li>
</ul>
</div>
<div class="section" id="id4">
<h2>[0.7.1] - 2019-06-08<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h2>
<p><strong>New features</strong></p>
<ul class="simple">
<li><p>Discrete Time Wright-Fisher simulation model (<a class="reference external" href="https://github.com/DomNelson">&#64;DomNelson</a>).</p></li>
<li><p>SMC/SMC’ simulation models (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>).</p></li>
<li><p>Mixed simulation models (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>).</p></li>
<li><p>Specify <code class="docutils literal notranslate"><span class="pre">end_time</span></code> to allow early-finish for simulations (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>).</p></li>
<li><p>Calculation of historical coalescence rates in the DemographyDebugger
(<a class="reference external" href="https://github.com/jgallowa07">&#64;jgallowa07</a>, <a class="reference external" href="https://github.com/petrelharp">&#64;petrelharp</a>).</p></li>
<li><p>Additional information on population sizes in DemographyDebugger
(<a class="reference external" href="https://github.com/andrewkern">&#64;andrewkern</a>).</p></li>
<li><p>Remove support for Python 2 (<a class="reference external" href="https://github.com/hugovk">&#64;hugovk</a>).</p></li>
<li><p>Allow specifying metadata for populations (<a class="reference external" href="https://github.com/jeromekelleher">&#64;jeromekelleher</a>).</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Various minor bug and doc fixes from <a class="reference external" href="https://github.com/hyanwong">&#64;hyanwong</a>, <a class="reference external" href="https://github.com/petrelharp">&#64;petrelharp</a>,
<a class="reference external" href="https://github.com/brianzhang01">&#64;brianzhang01</a>, <a class="reference external" href="https://github.com/mufernando">&#64;mufernando</a> and <a class="reference external" href="https://github.com/andrewkern">&#64;andrewkern</a>.</p></li>
</ul>
</div>
<div class="section" id="b1-2019-05-31">
<h2>[0.7.1b1] - 2019-05-31<a class="headerlink" href="#b1-2019-05-31" title="Permalink to this headline">¶</a></h2>
<p>Early release making DTWF code available to beta testers.</p>
</div>
<div class="section" id="id5">
<h2>[0.7.0] - 2019-02-19<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h2>
<p>Separation of tskit from msprime. Msprime is now solely dedicated to simulating
the coalescent, and all infrastucture for working with succinct tree sequences
is now provided by tskit. To ensure compatability, msprime now imports code
from tskit under the old names, which should ensure that all code continues
to work without changes.</p>
<p><strong>New features</strong></p>
<ul class="simple">
<li><p>Ability to record the full ARG (<a class="reference external" href="https://github.com/jerekoskela">&#64;jerekoskela</a>; <a class="reference external" href="https://github.com/tskit-dev/msprime/issues/665">#665</a>)</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Fix deprecation warning (<a class="reference external" href="https://github.com/tskit-dev/msprime/issues/695">#695</a>).</p></li>
</ul>
</div>
<div class="section" id="a1-2019-01-14">
<h2>[0.7.0a1] - 2019-01-14<a class="headerlink" href="#a1-2019-01-14" title="Permalink to this headline">¶</a></h2>
<p>Alpha release for testing the tskit/msprime split.</p>
</div>
<div class="section" id="id6">
<h2>[0.6.2] - 2018-12-04<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h2>
<p>Minor bugfix release.</p>
<p><strong>New features</strong>:
- Add provenance recording option to simplify (#601)
- Minor performance improvement (#598)</p>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Fix performance regression in replication (#608)</p></li>
</ul>
</div>
<div class="section" id="id7">
<h2>[0.6.1] - 2018-08-25<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h2>
<p>Significant features for integration with forwards-time simulators plus
improvements and bugfixes.</p>
<p><strong>Breaking changes</strong>:</p>
<ul class="simple">
<li><p>Change in the semantics of how populations are treated by simplify. By
default, populations that are not referenced will now be removed from the
data model. This can be avoided by setting <code class="docutils literal notranslate"><span class="pre">filter_populations=False</span></code>.</p></li>
<li><p>Simplify now raises an error if called on a set of tables that contain
one or more migrations.</p></li>
</ul>
<p><strong>New features</strong>:</p>
<ul class="simple">
<li><p>The simulate() function now supports a <code class="docutils literal notranslate"><span class="pre">from_ts</span></code> argument allowing
msprime to complete the ancestry in tree sequences generated by
forward simulations (#503, #541, #572, #581).</p></li>
<li><p>Add start_time and end_time parameters to the <code class="docutils literal notranslate"><span class="pre">mutate</span></code> function (#508).</p></li>
<li><p>Add <code class="docutils literal notranslate"><span class="pre">reduce_to_site_topology</span></code> argument to simplify. This allows us to
find the minimal tree sequence that would be visible from a given set
of sites, and is also a useful compression method if we are only interested
in the observed sequences. (#545, #307).</p></li>
<li><p>Simplify generalised to support individuals, and the <code class="docutils literal notranslate"><span class="pre">filter_populations</span></code>,
<code class="docutils literal notranslate"><span class="pre">filter_individuals</span></code> and <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code> parameters added to allow
filtering of unreferenced objects from the data model. (#567).</p></li>
<li><p>Default random seeds are now generated from a sequence initialised by
a system source of randomness (#534). Random seeds should also be safely generated
across multiple processes.</p></li>
<li><p>Full text I/0 support for Individuals and Populations (#498, #555)</p></li>
<li><p>Substantially improved performance in <code class="docutils literal notranslate"><span class="pre">msprime.load</span></code> for large tables
and significant refactoring of C code (#559, #567, #569).</p></li>
<li><p>Improved performance of generating genotypes (#580).</p></li>
<li><p>Formal schema for tree sequence provenance (#566, #583).</p></li>
<li><p>Many updates to documentation.</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Throw a more intelligle error during simulation if a topology is produced
where the time of a parent is equal to the time of the child. (#570, #87).</p></li>
<li><p>Pickle supported in the TableCollection object. (#574, #577).</p></li>
</ul>
<p><strong>Deprecated</strong>:</p>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">filter_zero_mutation_sites</span></code> parameter for simplify has been deprecated
in favour of <code class="docutils literal notranslate"><span class="pre">filter_sites</span></code>.</p></li>
</ul>
</div>
<div class="section" id="id8">
<h2>[0.6.0] - 2018-06-20<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h2>
<p>This release is focused on ensuring interoperability with the forthcoming SLiM
3.0 release, which has support for outputting tree sequences in msprime’s
.trees format. The release represents a substantial step towards the goal of
separating the <code class="docutils literal notranslate"><span class="pre">tskit</span></code> code from <code class="docutils literal notranslate"><span class="pre">msprime</span></code>. It removes the troublesome HDF5
dependency in favour of the much simpler <code class="docutils literal notranslate"><span class="pre">kastore</span></code> library.</p>
<p>The principle new features are the mutate() function which allows us to easily
add mutations to any tree sequence, preliminary support for Individuals and
Populations within the data model, and the addition of the new TableCollection
object as the central structure in the Tables API.</p>
<p><strong>Breaking changes</strong>:</p>
<ul class="simple">
<li><p>Files stored in the HDF5 format will need to upgraded using the
<code class="docutils literal notranslate"><span class="pre">msp</span> <span class="pre">upgrade</span></code> command.</p></li>
</ul>
<p><strong>New features</strong>:</p>
<ul class="simple">
<li><p>The mutate function (#507).</p></li>
<li><p>Removed HDF5 library dependency. Now use the embedded kastore library
for storing data.</p></li>
<li><p>Numpy and h5py are now install time dependencies, solving some installation
headaches.</p></li>
<li><p>The new TableCollection type  gives much tighter integration with the
low-level library. Functions like sort_tables and simplify_tables are
now methods of this class. The load_tables function has been replaced
by TableCollection.tree_sequence. These functions still work, but are
deprecated.</p></li>
<li><p>Preliminary support for Individual and Population types in the Tables
API and for TreeSequences.</p></li>
<li><p>Add ‘root’ argument to SparseTree.newick and support for arbitrary
node labels (#510).</p></li>
<li><p>Larger numbers of alleles now supported via 16-bit genotypes (#466).</p></li>
<li><p>Substantially improved simplify performance when there is a large
number of sites (#453).</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Fix bug in tree drawing with many roots (#486)</p></li>
<li><p>Fix segfault in accessing trees with zero roots (#515)</p></li>
<li><p>Fix bug where DemographyDebugger was modifying the input sample sizes (#407)</p></li>
</ul>
<p><strong>Deprecated</strong>:</p>
<ul class="simple">
<li><p>sort_tables is deprecated in favour of TableCollection.sort().</p></li>
<li><p>simplify_tables is deprecated in favour of TableCollection.simplify().</p></li>
<li><p>load_tables is deprecated in favour of TableCollection.tree_sequence().</p></li>
</ul>
</div>
<div class="section" id="id9">
<h2>[0.5.0] - 2018-02-26<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h2>
<p>This is a major update to the underlying data structures in msprime to
generalise the information that can be modelled, and allow
for data from external sources to be efficiently processed. The
new Tables API enables efficient interchange of tree sequence data using
numpy arrays. Many updates have also been made to the tree sequence
API to make it more Pythonic and general. Most changes are backwards
compatible, however.</p>
<p><strong>Breaking changes</strong>:</p>
<ul>
<li><p>The <code class="docutils literal notranslate"><span class="pre">SparseTree.mutations()</span></code> and <code class="docutils literal notranslate"><span class="pre">TreeSequence.mutations()</span></code> iterators no
longer support tuple-like access to values. For example, code like</p>
<blockquote>
<div><dl class="simple">
<dt>for x, u, j in ts.mutations():</dt><dd><p>print(“mutation at position”, x, “node = “, u)</p>
</dd>
</dl>
</div></blockquote>
<p>will no longer work. Code using the old <code class="docutils literal notranslate"><span class="pre">Mutation.position</span></code> and
<code class="docutils literal notranslate"><span class="pre">Mutation.index</span></code> will still work through deprecated aliases,
but new code should access these values through <code class="docutils literal notranslate"><span class="pre">Site.position</span></code>
and <code class="docutils literal notranslate"><span class="pre">Site.id</span></code>, respectively.</p>
</li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">TreeSequence.diffs()</span></code> method no longer works. Please use
the <code class="docutils literal notranslate"><span class="pre">TreeSequence.edge_diffs()</span></code> method instead.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">TreeSequence.get_num_records()</span></code> no longer works. Any code using
this or the <code class="docutils literal notranslate"><span class="pre">records()</span></code> iterator should be rewritten to work with
the <code class="docutils literal notranslate"><span class="pre">edges()</span></code> iterator and num_edges instead.</p></li>
<li><p>Files stored in the HDF5 format will need to upgraded using the
<code class="docutils literal notranslate"><span class="pre">msp</span> <span class="pre">upgrade</span></code> command.</p></li>
</ul>
<p><strong>New features</strong>:</p>
<ul class="simple">
<li><p>The API has been made more Pythonic by replacing (e.g.)
<code class="docutils literal notranslate"><span class="pre">tree.get_parent(u)</span></code> with <code class="docutils literal notranslate"><span class="pre">tree.parent(u)</span></code>, and
<code class="docutils literal notranslate"><span class="pre">tree.get_total_branch_length()</span></code> with <code class="docutils literal notranslate"><span class="pre">tree.total_branch_length</span></code>.
The old forms have been maintained as deprecated aliases. (#64)</p></li>
<li><p>Efficient interchange of tree sequence data using the new Tables
API. This consists of classes representing the various
tables (e.g. <code class="docutils literal notranslate"><span class="pre">NodeTable</span></code>) and some utility functions (such
as <code class="docutils literal notranslate"><span class="pre">load_tables</span></code>, <code class="docutils literal notranslate"><span class="pre">sort_tables</span></code>, etc).</p></li>
<li><p>Support for a much more general class of tree sequence topologies.
For example, trees with multiple roots are fully supported.</p></li>
<li><p>Substantially generalised mutation model. Mutations now occur at
specific sites, which can be associated with zero to many mutations.
Each site has an ancestral state (any character string) and
each mutation a derived state (any character string).</p></li>
<li><p>Substantially updated documentation to rigorously define the
underlying data model and requirements for imported data.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">variants()</span></code> method now returns a list of alleles for each
site, and genotypes are indexes into this array. This is both
consistent with existing usage and works with the newly generalised
mutation model, which allows arbitrary strings of characters as
mutational states.</p></li>
<li><p>Add the formal concept of a sample, and distinguished from ‘leaves’.
Change <code class="docutils literal notranslate"><span class="pre">tracked_leaves</span></code>, etc. to <code class="docutils literal notranslate"><span class="pre">tracked_samples</span></code> (#225).
Also rename <code class="docutils literal notranslate"><span class="pre">sample_size</span></code> to <code class="docutils literal notranslate"><span class="pre">num_samples</span></code> for consistency (#227).</p></li>
<li><p>The simplify() method returns subsets of a large tree sequence.</p></li>
<li><p>TreeSequence.first() returns the first tree in sequence.</p></li>
<li><p>Windows support. Msprime is now routinely tested on Windows as
part of the suite of continuous integration tests.</p></li>
<li><p>Newick output is not supported for more general trees. (#117)</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">genotype_matrix</span></code> method allows efficient access to the
full genotype matrix. (#306)</p></li>
<li><p>The variants iterator no longer uses a single buffer for
genotype data, removing a common source of error (#253).</p></li>
<li><p>Unicode and ASCII output formats for <code class="docutils literal notranslate"><span class="pre">SparseTree.draw()</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SparseTree.draw()</span></code> renders tree in the more conventional ‘square
shoulders’ format.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SparseTree.draw()</span></code> by default returns an SVG string, so it can
be easily displayed in a Jupyter notebook. (#204)</p></li>
<li><p>Preliminary support for a broad class of site-based statistics,
including Patterson’s f-statistics, has been added, through
the <cite>SiteStatCalculator</cite>, and its branch length analog,
<cite>BranchLengthStatCalculator</cite>.  The interface is still in development,
and is expected may change.</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Duplicate site no longer possible (#159)</p></li>
<li><p>Fix for incorrect population sizes in DemographyDebugger (#66).</p></li>
</ul>
<p><strong>Deprecated</strong>:</p>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">records</span></code> iterator has been deprecated, and the underlying data
model has moved away from the concept of coalescence records. The
structure of a tree sequence is now defined in terms of a set of nodes
and edges, essentially a normlised version of coalescence records.</p></li>
<li><p>Changed <code class="docutils literal notranslate"><span class="pre">population_id</span></code> to <code class="docutils literal notranslate"><span class="pre">population</span></code> in various DemographicEvent
classes for consistency. The old <code class="docutils literal notranslate"><span class="pre">population_id</span></code> argument is kept as a
deprecated alias.</p></li>
<li><p>Changed <code class="docutils literal notranslate"><span class="pre">destination</span></code> to <code class="docutils literal notranslate"><span class="pre">dest</span></code> in MassMigrationEvent. The old
<code class="docutils literal notranslate"><span class="pre">destination</span></code> argument is retained as a deprecated alias.</p></li>
<li><p>Changed <code class="docutils literal notranslate"><span class="pre">sample_size</span></code> to <code class="docutils literal notranslate"><span class="pre">num_samples</span></code> in TreeSequence and
SparseTree. The older versions are retained as deprecated aliases.</p></li>
<li><p>Change <code class="docutils literal notranslate"><span class="pre">get_num_leaves</span></code> to <code class="docutils literal notranslate"><span class="pre">num_samples</span></code> in SparseTree. The
<code class="docutils literal notranslate"><span class="pre">get_num_leaves</span></code> method (and other related methods) that have
been retained for backwards compatability are semantically incorrect,
in that they now return the number of <strong>samples</strong>. This should have
no effect on existing code, since samples and leaves were synonymous.
New code should use the documented <code class="docutils literal notranslate"><span class="pre">num_samples</span></code> form.</p></li>
<li><p>Accessing the <code class="docutils literal notranslate"><span class="pre">position</span></code> attribute on a <code class="docutils literal notranslate"><span class="pre">Mutation</span></code> or
<code class="docutils literal notranslate"><span class="pre">Variant</span></code> object is now deprecated, as this is a property of a <code class="docutils literal notranslate"><span class="pre">Site</span></code>.</p></li>
<li><p>Accessing the <code class="docutils literal notranslate"><span class="pre">index</span></code> attribute on a <code class="docutils literal notranslate"><span class="pre">Mutation</span></code> or <code class="docutils literal notranslate"><span class="pre">Variant</span></code> object
is now deprecated. Please use <code class="docutils literal notranslate"><span class="pre">variant.site.id</span></code> instead. In general,
objects with IDs (i.e., derived from tables) now have an <code class="docutils literal notranslate"><span class="pre">id</span></code> field.</p></li>
<li><p>Various <code class="docutils literal notranslate"><span class="pre">get_</span></code> methods in TreeSequence and SparseTree have been
replaced by more Pythonic alternatives.</p></li>
</ul>
</div>
<div class="section" id="id10">
<h2>[0.4.0] - 2016-10-16<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h2>
<p>Major release providing new functionality and laying groundwork for
upcoming functionality.</p>
<p><strong>Breaking changes</strong>:</p>
<ul>
<li><p>The HDF5 file format has been changed to allow for non-binary trees
and to improve performance. It is now both smaller and faster to
load. However, msprime cannot directly load tree sequence files
written by older versions. The <code class="docutils literal notranslate"><span class="pre">msp</span> <span class="pre">upgrade</span></code> utility has been
developed to provide an upgrade path for existing users, so that
files written by older versions of msprime can be converted to the
newer format and read by version 0.4.x of msprime.</p></li>
<li><p>The tuples returned by the <code class="docutils literal notranslate"><span class="pre">mutations</span></code> method contains an element.
This will break code doing things like</p>
<blockquote>
<div><dl class="simple">
<dt>for pos, node in ts.mutations():</dt><dd><p>print(pos, node)</p>
</dd>
</dl>
</div></blockquote>
<p>For better forward compatibility, code should use named attributes
rather than positional access:</p>
<blockquote>
<div><dl class="simple">
<dt>for mutation in ts.mutations():</dt><dd><p>print(mutation.position, mutation.node)</p>
</dd>
</dl>
</div></blockquote>
</li>
<li><p>Similarly, the undocumented <code class="docutils literal notranslate"><span class="pre">variants</span></code> method has some major changes:</p>
<ol class="arabic simple">
<li><p>The returned tuple has two new values, <code class="docutils literal notranslate"><span class="pre">node</span></code> and <code class="docutils literal notranslate"><span class="pre">index</span></code>
in the middle of the tuple (but see the point above about using
named attributes).</p></li>
<li><p>The returned genotypes are by default numpy arrays. To revert
to the old behaviour of returning Python bytes objects, use the
<code class="docutils literal notranslate"><span class="pre">as_bytes</span></code> argument to the <code class="docutils literal notranslate"><span class="pre">variants()</span></code> method.</p></li>
</ol>
</li>
</ul>
<p><strong>New features</strong>:</p>
<ul class="simple">
<li><p>Historical samples. Using the <code class="docutils literal notranslate"><span class="pre">samples</span></code> argument to <code class="docutils literal notranslate"><span class="pre">simulate</span></code>
users can specify the location and time of all samples explicitly.</p></li>
<li><p>HDF5 file upgrade utility <code class="docutils literal notranslate"><span class="pre">msp</span> <span class="pre">upgrade</span></code></p></li>
<li><p>Support for non-binary trees in the tree sequence, and relaxation
of the requirements on input tree sequences using the read_txt()
function.</p></li>
<li><p>Integration with numpy, with zero-copy access to the low-level C API.</p></li>
<li><p>Documented the variants() method that provides access to the sample
genotypes as either numpy arrays or Python bytes objects.</p></li>
<li><p>New LdCalculator class that allows very fast calculation of r^2 values.</p></li>
<li><p>Initial support for threading.</p></li>
<li><p>The values returned mutations() method now also contain an <code class="docutils literal notranslate"><span class="pre">index</span></code>
attribute. This makes many operations simpler.</p></li>
<li><p>New TreeSequence.get_time() method that returns the time a sample
was sampled at.</p></li>
</ul>
<p><strong>Performance improvements</strong>:</p>
<ul class="simple">
<li><p>File load times substantially reduced by pre-computing and storing
traversal indexes.</p></li>
<li><p>O(1) implementation of TreeSequence.get_num_trees()</p></li>
<li><p>Improved control of enabled tree features in TreeSequence.trees()
method using the <code class="docutils literal notranslate"><span class="pre">leaf_lists</span></code> and <code class="docutils literal notranslate"><span class="pre">leaf_counts</span></code> arguments.</p></li>
</ul>
<p><strong>Bug fixes</strong>:</p>
<ul class="simple">
<li><p>Fixed a precision problem with DemographyDebugger. #37</p></li>
<li><p>Segfault on large haplotypes. #29</p></li>
</ul>
</div>
<div class="section" id="id11">
<h2>[0.3.2] - 2016-07-21<a class="headerlink" href="#id11" title="Permalink to this headline">¶</a></h2>
<p>Feature release adding new import and export features to the API
and CLI.</p>
<ul class="simple">
<li><p>New <code class="docutils literal notranslate"><span class="pre">TreeSequence.write_records</span></code> and <code class="docutils literal notranslate"><span class="pre">TreeSequence.write_mutations</span></code>
methods to serialise a tree sequence in a human readable text format.</p></li>
<li><p>New <code class="docutils literal notranslate"><span class="pre">msprime.load_txt()</span></code> method that parses the above formats, and
allows msprime to read in data from external sources.</p></li>
<li><p>New <code class="docutils literal notranslate"><span class="pre">TreeSequence.write_vcf</span></code> method to write mutation information
in VCF format.</p></li>
<li><p>Miscellaneous documentation fixes.</p></li>
</ul>
</div>
<div class="section" id="id12">
<h2>[0.3.1] - 2016-06-24<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h2>
<p>Feature release adding population related methods to the API.</p>
<ul class="simple">
<li><p>New <code class="docutils literal notranslate"><span class="pre">TreeSequence.get_population(sample_id)</span></code> method.</p></li>
<li><p>New <code class="docutils literal notranslate"><span class="pre">TreeSequence.get_samples(population_id)</span></code> method.</p></li>
<li><p>Added the optional <code class="docutils literal notranslate"><span class="pre">samples</span></code> argument to the
<code class="docutils literal notranslate"><span class="pre">TreeSequence.get_pairwise_diversity</span></code> method.</p></li>
<li><p>Fixed a potential low-level buffer overrun problem.</p></li>
</ul>
</div>
<div class="section" id="id13">
<h2>[0.3.0] - 2016-05-31<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h2>
<p>Bugfix release affecting all users of the Python API. Version 0.2.0 contained a
confusing and inconsistent mix of times and rates being expressed in both
coalescent units and generations. This release changes _all_ times and rates
used when describing demographic models to generations, and also changes
all population sizes to be absolute. In the interest of consistency, the
units of the trees output by msprime are also changed to generations. This
is a major breaking change, and will require updates to all scripts using the
API.</p>
<p>This release also include some performance improvements and additional
functionality.</p>
<p>Mspms users are not affected, other than benefiting from performance
improvements.</p>
<p>Breaking changes:</p>
<ul class="simple">
<li><p>Time values are now rescaled into generations when a TreeSequence is
created, and so all times associated with tree nodes are measured in
generations. The time values in any existing HDF5 file will now be
interpreted as being in generations, so stored simulations must be
rerun. To minimise the chance of this happening silently, we have
incremented the file format major version number, so that attempts
to read older versions will fail.</p></li>
<li><p>Growth rate values for the PopulationConfiguration class are now
per generation, and population sizes are absolute. These were in
coalescent units and relative to Ne previously.</p></li>
<li><p>GrowthRateChangeEvents and SizeChangeEvents have been replaced with
a single class, PopulationParametersChange. This new class takes
an initial_size as the absolute population size, and growth_rate
per generation. Since the change in units was a breaking one,
potentially leading to subtle and confusing bugs, we decided that
the name refactoring would at least ensure that users would need
to be aware that the change had been made. This API should now
be stable, and will not be changed again without an excellent
reason.</p></li>
<li><p>MigrationRateChangeEvent has been renamed to MigrationRateChange
and the migration rates are now per-generation.</p></li>
<li><p>MassMigrationEvent has been renamed to MassMigration, and the
values of source and destination swapped, fixing the bug in
issue #14.</p></li>
<li><p>The TreeSequence.records() method now returns an extra value,
potentially breaking client code.</p></li>
</ul>
<p>Improvements:</p>
<ul class="simple">
<li><p>Added tutorial for demographic events.</p></li>
<li><p>Added DemographyDebugger class to help view the changes in populations
over time.</p></li>
<li><p>Added population tracking for coalescent events. We can now determine
the population associated with every tree node. The relevant information
has been added to the HDF5 file format.</p></li>
<li><p>Improved performance for replication by reusing the same low-level
simulator instance. This leads to significant improvements for large
numbers of replicates of small simulations. Issue #8.</p></li>
<li><p>Changed the TreeSequence.records() method to return named tuples.</p></li>
<li><p>Added get_total_branch_length method. Issue #12.</p></li>
<li><p>Fixed bug in reading Hapmap files. Issue #13.</p></li>
</ul>
</div>
<div class="section" id="id14">
<h2>[0.2.0] - 2016-05-05<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h2>
<p>Major update release, adding significant new functionality to the Python
API and several breaking changes. All code written for the 0.1.x API
will be affected, unfortunately.</p>
<p>Breaking changes:</p>
<ul class="simple">
<li><p>Sample IDs are now zero indexed. In previous versions of msprime, the
samples were numbered from 1 to n inclusive, which is not Pythonic.
This change has been made to make the API more usable, but will
cause issues for existing code.</p></li>
<li><p>There is now an Ne parameter to simulate(), and recombination,
mutation and migration rates are now all per-generation. The
keyword arguments have been changed to recombination_rate
and mutation_rate, which should mean that silent errors will
be avoided. All rates in existing code will need to be
divided by 4 as a result of this. This change was made to make
working with recombination maps and per generation recombination
rates easier.</p></li>
<li><p>Msprime now uses continuous values to represent coordinates, and
the num_loci parameter has been replaced with a new length parameter
to simulate(). Internally, a discrete recombination model is still
used, but by default the potential number of discrete sites is
very large and effectively continuous. True discrete recombination
models can still be specified by using the recombination_map
argument to simulate.</p></li>
<li><p>The population_models argument to simulate() has been removed, and
replaced with the population_configuration and demographic_events
parameters. This was necessary to provide the full demographic
model.</p></li>
<li><p>The HDF5 file format has been updated to accommodate the continuous
coordinates, along with other minor changes. As a consequence,
simulation results will be somewhat larger. Stored simulations will
need to be re-run and saved.</p></li>
<li><p>Removed the random_seed key from the provenance JSON strings.</p></li>
<li><p>Removed the simulate_tree() function, as it seemed to offer little
extra value.</p></li>
</ul>
<p>New features:</p>
<ul class="simple">
<li><p>Simulation of variable recombination rates via arbitrary recombination
maps.</p></li>
<li><p>Full support for population structure and demographic events.</p></li>
<li><p>API support for replication via the num_replicates argument to simulate().</p></li>
<li><p>Fully reworked random generation mechanisms, so that in the nominal
case a single instance of gsl_rng is used throughout the entire
simulation session.</p></li>
<li><p>Addition of several miscellaneous methods to the TreeSequence API.</p></li>
<li><p>Added NULL_NODE constant to make tree traversals more readable.</p></li>
</ul>
</div>
<div class="section" id="id15">
<h2>[0.1.10] - 2016-04-21<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h2>
<p>Bugfix release. Fixes serious issue affecting simulations with small
sample sizes.</p>
<p><a class="reference external" href="https://github.com/jeromekelleher/msprime/issues/7">https://github.com/jeromekelleher/msprime/issues/7</a></p>
<p>All users of mspms should update immediately and any analyses using
a small sample size (&lt; 10) with mutations should be repeated.</p>
<p>Many thanks to Konrad Lohse for identifying the issue.</p>
</div>
<div class="section" id="id16">
<h2>[0.1.9] - 2016-04-01<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h2>
<p>Bugfix release. Fixes serious issue affecting random seeds in mspms.</p>
<p><a class="reference external" href="https://github.com/jeromekelleher/msprime/issues/6">https://github.com/jeromekelleher/msprime/issues/6</a></p>
<p>All users of mspms should update immediately and any analyses using
the <code class="docutils literal notranslate"><span class="pre">-seeds</span></code> option in mspms should be repeated.</p>
<p>Many thanks to Derek Setter for identifying the issue.</p>
</div>
<div class="section" id="id17">
<h2>[0.1.8] - 2016-02-17<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h2>
<p>Transitional release providing population structure support for the
ms-compatible command line interface. A considerable amount of low-level
plumbing was required to provide the required flexibility. This is currently
not visible from the high-level API, but will shortly be made available in the
forthcoming 0.2.x series.</p>
<p>The current implementation of migration should work well for small numbers of
populations (e.g. &lt; 10), but will not scale well for large numbers of
populations.</p>
<div class="section" id="changes">
<h3>Changes<a class="headerlink" href="#changes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Added the -I, -m, -ma, -em, -eM, -ema, -eG, -eg, -eN, -en,
-ej and -es options to mspms. These should provide full ms
compatability, except for the -es option which is currently
limited in scope.</p></li>
<li><p>Added some extra keys to the low-level configuration JSON in
the HDF5 file format to describe the population structure.
This will be documented in a future release.</p></li>
<li><p>Added a <cite>get_pairwise_diversity</cite> method to the TreeSequence
class to efficiently calculate the population genetics
statistic pi.</p></li>
</ul>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="tutorial.html" class="btn btn-neutral float-right" title="PORTING IN PROGRESS: Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="CITATION.html" class="btn btn-neutral float-left" title="Citing msprime" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2015-2020, Tskit Developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>