

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Demography &mdash; msprime 0.1.dev1 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/jupyter-sphinx.css" type="text/css" />
  <link rel="stylesheet" href="_static/thebelab.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script src="_static/thebelab-helper.js"></script>
        <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
        <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Likelihoods" href="likelihoods.html" />
    <link rel="prev" title="Mutations" href="mutations.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> msprime
          

          
          </a>

          
            
            
              <div class="version">
                0.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart</a></li>
<li class="toctree-l1"><a class="reference internal" href="ancestry.html">Ancestry</a></li>
<li class="toctree-l1"><a class="reference internal" href="mutations.html">Mutations</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Demography</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#old-text">Old text</a></li>
<li class="toctree-l2"><a class="reference internal" href="#api-documentation">API Documentation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#populations">Populations</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#deprecated-api">Deprecated API</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#migration-rates">Migration rates</a></li>
<li class="toctree-l2"><a class="reference internal" href="#demographic-events">Demographic Events</a></li>
<li class="toctree-l2"><a class="reference internal" href="#debugging-demographic-models">Debugging demographic models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#species-trees">Species trees</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="likelihoods.html">Likelihoods</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickref.html">API quick reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="utilities.html">Utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command line interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="development.html">Developer documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="CITATION.html">Citing msprime</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">PORTING IN PROGRESS: Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">PORTING IN PROGRESS: API</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">msprime</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Demography</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/demography.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="demography">
<span id="sec-demography"></span><h1>Demography<a class="headerlink" href="#demography" title="Permalink to this headline">¶</a></h1>
<div class="admonition-todo admonition" id="id1">
<p class="admonition-title">Todo</p>
<p>This section needs some intro material</p>
</div>
<div class="section" id="old-text">
<h2>Old text<a class="headerlink" href="#old-text" title="Permalink to this headline">¶</a></h2>
<div class="admonition-todo admonition" id="id2">
<p class="admonition-title">Todo</p>
<p>This is the text from the old API docs. Reuse somewhere more appropriate,
either by having a “summary” section of by moving bits of the text into the
relevant sections.</p>
</div>
<p>Population structure is modelled by specifying a fixed number of subpopulations
<span class="math notranslate nohighlight">\(d\)</span>, and a <span class="math notranslate nohighlight">\(d \times d\)</span> matrix <span class="math notranslate nohighlight">\(M\)</span> of per-generation
migration rates. The <span class="math notranslate nohighlight">\((j,k)^{th}\)</span> entry of <span class="math notranslate nohighlight">\(M\)</span> is the expected number
of migrants moving from population <span class="math notranslate nohighlight">\(k\)</span> to population <span class="math notranslate nohighlight">\(j\)</span> per
generation, divided by the size of population <span class="math notranslate nohighlight">\(j\)</span>. In terms of the
coalescent process, <span class="math notranslate nohighlight">\(M_{j,k}\)</span> gives the rate at which an ancestral
lineage moves from population <span class="math notranslate nohighlight">\(j\)</span> to population <span class="math notranslate nohighlight">\(k\)</span>, as one follows
it back through time. In continuous-time models, when <span class="math notranslate nohighlight">\(M_{j,k}\)</span> is close
to zero, this rate is approximately equivalent to the fraction of population <span class="math notranslate nohighlight">\(j\)</span>
that is replaced each generation by migrants from population <span class="math notranslate nohighlight">\(k\)</span>. In
discrete-time models, the equivalence is exact and each row of <span class="math notranslate nohighlight">\(M\)</span> has
the constraint <span class="math notranslate nohighlight">\(\sum_{k \neq j} M_{j,k} \leq 1\)</span>. This differs from the
migration matrix one usually uses in population demography: if <span class="math notranslate nohighlight">\(m_{k,j}\)</span>
is the proportion of individuals (in the usual sense; not lineages) in
population <span class="math notranslate nohighlight">\(k\)</span> that move to population <span class="math notranslate nohighlight">\(j\)</span> per generation, then
translating this proportion of population <span class="math notranslate nohighlight">\(k\)</span> to a proportion of
population <span class="math notranslate nohighlight">\(j\)</span>, we have <span class="math notranslate nohighlight">\(M_{j,k} = m_{k,j} \times N_k / N_j\)</span>.</p>
<p>Each subpopulation has an initial absolute population size <span class="math notranslate nohighlight">\(s\)</span>
and a per generation exponential growth rate <span class="math notranslate nohighlight">\(\alpha\)</span>. The size of a
given population at time <span class="math notranslate nohighlight">\(t\)</span> in the past (measured in generations) is
therefore given by <span class="math notranslate nohighlight">\(s e^{-\alpha t}\)</span>. Demographic events that occur in
the history of the simulated population alter some aspect of this population
configuration at a particular time in the past.</p>
</div>
<div class="section" id="api-documentation">
<h2>API Documentation<a class="headerlink" href="#api-documentation" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="msprime.Demography">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">Demography</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">populations</span><span class="o">=</span><span class="default_value">NOTHING</span></em>, <em class="sig-param"><span class="n">migration_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">events</span><span class="o">=</span><span class="default_value">NOTHING</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography" title="Permalink to this definition">¶</a></dt>
<dd><p>A description of a demographic model for an msprime simulation.</p>
<p>TODO document properly.</p>
<dl class="py method">
<dt id="msprime.Demography.from_old_style">
<em class="property">static </em><code class="sig-name descname">from_old_style</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">population_configurations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">migration_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">demographic_events</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.from_old_style"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.from_old_style" title="Permalink to this definition">¶</a></dt>
<dd><p>Creates a Demography object from the pre 1.0 style input parameters,
reproducing the old semantics with respect to default values.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.insert_populations">
<code class="sig-name descname">insert_populations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">tables</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.insert_populations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.insert_populations" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert population definitions for this demography into the specified
set of tables.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.island_model">
<em class="property">static </em><code class="sig-name descname">island_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_populations</span></em>, <em class="sig-param"><span class="n">migration_rate</span></em>, <em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.island_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.island_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object with the specified number of
populations with symmetric migration between all pairs of populations
at the specified rate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of populations in this Island
model.</p></li>
<li><p><strong>migration_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The migration rate between each pair of
populations.</p></li>
<li><p><strong>Ne</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The initial size of each population. Defaults to 1.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Demography object representing this model, suitable as
input to <a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.sample">
<code class="sig-name descname">sample</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.sample"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.sample" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list of <code class="xref py py-class docutils literal notranslate"><span class="pre">Sample</span></code> objects, with the number of samples
from each population determined by either the positional or keyword
arguments. The keyword and positional forms cannot be mixed.</p>
<p>Positional arguments are intepreted as the number of samples to draw
from the corresponding population index. For example,
<code class="docutils literal notranslate"><span class="pre">demography.sample.(2,</span> <span class="pre">5,</span> <span class="pre">7)</span></code> returns a list of 14 samples, two of
which are from the model’s first population, five from the  second
population, and seven from the third. The number of positional
arguments must be less than or equal to the number of populations. If
the number of arguments is less than the number of populations, then
remaining samples sizes are treated as zero. For example, in a two
population model, <code class="docutils literal notranslate"><span class="pre">demography.sample(5)</span></code> is equivalent to
<code class="docutils literal notranslate"><span class="pre">demography.sample(5,</span> <span class="pre">0)</span></code>.</p>
<p>Samples can also be generated based on the population’s name
using keyword arguments. For example, if we have a demographic
model with three populations named “X”, “Y”, and “Z” we can
use <code class="docutils literal notranslate"><span class="pre">demography.sample(X=1,</span> <span class="pre">Y=2,</span> <span class="pre">Z=3)</span></code> will return six samples:
one for the population named “X”, two for “Y” and three for “Z”.</p>
<p>The list of <code class="xref py py-class docutils literal notranslate"><span class="pre">Sample</span></code> objects returned is always grouped
by population. In the positional case, population IDs in
nondecreasing numerical order. When the keyword form is used,
samples are returned in the order in which they were specified.
Thus, the list returned by <code class="docutils literal notranslate"><span class="pre">demography.sample(A=1,</span> <span class="pre">B=2)</span></code>
will have one sample for population “A”, followed by two samples
for population “B”. This is true regardless of the numerical
population IDs for these populations.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>A list of <code class="xref py py-class docutils literal notranslate"><span class="pre">Sample</span></code> instances.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p><a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>(Sample)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.simple_model">
<code class="sig-name descname">simple_model</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.simple_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.simple_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a simple single-population model.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.stepping_stone_1d">
<em class="property">static </em><code class="sig-name descname">stepping_stone_1d</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_populations</span></em>, <em class="sig-param"><span class="n">migration_rate</span></em>, <em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">circular</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.stepping_stone_1d"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.stepping_stone_1d" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a> object describing a 1 dimensional
stepping stone model where adjacent populations exchange migrants at
the specified rate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_populations</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of populations in this Island
model.</p></li>
<li><p><strong>migration_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The migration rate between each pair of
populations.</p></li>
<li><p><strong>Ne</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The initial size of each of the populations. Defaults
to 1.</p></li>
<li><p><strong>circular</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – If True (the default) the stepping stone model is
circular, so that migration occurs between the first and last
populations. If False, no migration occurs.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A Demography object representing this model, suitable as
input to <a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#msprime.Demography" title="msprime.Demography">Demography</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.Demography.validate">
<code class="sig-name descname">validate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Demography.validate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Demography.validate" title="Permalink to this definition">¶</a></dt>
<dd><p>Checks the demography looks sensible and raises errors/warnings
appropriately.</p>
</dd></dl>

</dd></dl>

<div class="admonition-todo admonition" id="id3">
<p class="admonition-title">Todo</p>
<p>These functions should be static methods of Demography.</p>
</div>
<dl class="py function">
<dt id="msprime.parse_species_tree">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">parse_species_tree</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/species_trees.html#parse_species_tree"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.parse_species_tree" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses a species tree in
<a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick</a> format and defines a
simulation model according to the species tree. The tree is assumed to be
rooted and ultrametric and branch lengths must be included and correspond to
time, either in units of millions of years, years, or generations. Leaves must
be named.</p>
<p>After reading the input tree, this function defines a
<a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instance for each terminal node in the tree,
corresponding to extant species. These population configurations store the
species’ name and population size. The specified Ne is used as the size of all
populations. Additionally, one or more <a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> instances are
defined for each internal node, with the time of the mass migration set
according to the age of the node in the species tree. <span class="math notranslate nohighlight">\(n - 1\)</span> mass
migration events are defined for internal nodes with <cite>n</cite> descendants, meaning
that a single event is defined for bifurcating nodes. For each internal node,
the left-most of the descendant populations is arbitrarily selected as the
destination in all mass migrations defined for that node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The tree string in Newick format, with named leaves and branch
lengths.</p></li>
<li><p><strong>Ne</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The effective population size.</p></li>
<li><p><strong>branch_length_units</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The units of time in which the species tree’s
branch lengths are measured. Allowed branch length units are millions of
years, years, and generations; these should be specified with the strings
<code class="docutils literal notranslate"><span class="pre">&quot;myr&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;yr&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;gen&quot;</span></code>, respectively. This defaults to
<code class="docutils literal notranslate"><span class="pre">&quot;gen&quot;</span></code>.</p></li>
<li><p><strong>generation_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a><em> or </em><a class="reference external" href="https://docs.python.org/3/library/constants.html#None" title="(in Python v3.9)"><em>None</em></a>) – The number of years per generation. If and only
if the branch lengths are not in units of generations, the generation time
must be specified. This defaults to <cite>None</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of two lists of which the first contains
<a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instances and the second contains
<a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> instances. The population configurations specify
the size of each population and the species name corresponding to each
population. Species names are stored as metadata in each
<a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instance, with the metadata tag
“species_name”. Sampling configurations and growth rates are not specified
in the population configurations. The list of population configurations is
ordered according to the order of the corresponding extant species in a
<a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Post-order_(LRN)">post-order tree traversal</a>. The list
of mass migration events is ordered by the time of the events, from young
to old events.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>)</p>
</dd>
<dt class="field-even">Warning</dt>
<dd class="field-even"><p>This function does not modify migration matrices. When the population
configurations and mass migration events returned by this function are used
to simulate with the <a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a> function, it should be ensured that
migration rates to source populations of mass migration events are zero
after the mass migration (viewed backwards in time).</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="msprime.parse_starbeast">
<code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">parse_starbeast</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/species_trees.html#parse_starbeast"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.parse_starbeast" title="Permalink to this definition">¶</a></dt>
<dd><p>This function parses a species tree produced by the program <a class="reference external" href="https://www.beast2.org/treeannotator">TreeAnnotator</a>
based on a posterior tree distribution generated with <a class="reference external" href="https://academic.oup.com/mbe/article/34/8/2101/3738283">StarBEAST</a>  and defines a
simulation model according to the species tree. Species trees produced by
TreeAnnotator are written in <a class="reference external" href="https://en.wikipedia.org/wiki/Nexus_file">Nexus</a> format and are rooted,
bifurcating, and ultrametric. Branch lengths usually are in units of millions
of years, but the use of other units is permitted by StarBEAST (and thus
TreeAnnotator). This function allows branch length units of millions of years
or years. Leaves must be named and the tree must include information on
population sizes of leaf and ancestral species in the form of annotation with
the “dmv” tag, which is the case for trees written by TreeAnnotator based on
StarBEAST posterior tree distributions.</p>
<p>After reading the input tree, this function defines a
<a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instance for each terminal node in the tree,
corresponding to extant species. These population configurations store the
species’ name and population size, both according to information from the input
tree. Additionally, a <a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> instance is defined for each
internal node, with the time of the mass migration set according to the age of
the node in the species tree. For each internal node, the left one of the two
descendant populations is arbitrarily selected as the destination in the mass
migration defined for that node. A <a class="reference internal" href="#msprime.PopulationParametersChange" title="msprime.PopulationParametersChange"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationParametersChange</span></code></a>
instance is also added for each internal node to adjust the population
size of the destination population according to the information given in the
tree for the population size of the species that is ancestral to the node. Like
the mass migration event defined for the same node, the time of the population
parameter change is also set according to the age of the node.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tree</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The tree string in Nexus format, with named leaves, branch
lengths, and branch annotation. Typically, this string is the entire content
of a file written by TreeAnnotator.</p></li>
<li><p><strong>generation_time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The number of years per generation.</p></li>
<li><p><strong>branch_length_units</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The units of time in which the species tree’s
branch lengths are measured. Allowed branch length units are millions of
years, and years; these should be specified with the strings <code class="docutils literal notranslate"><span class="pre">&quot;myr&quot;</span></code> or
<code class="docutils literal notranslate"><span class="pre">&quot;yr&quot;</span></code>, respectively. This defaults to <code class="docutils literal notranslate"><span class="pre">&quot;myr&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><p>A tuple of two lists of which the first contains
<a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instances and the second contains
<a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a> and <a class="reference internal" href="#msprime.PopulationParametersChange" title="msprime.PopulationParametersChange"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationParametersChange</span></code></a> instances.
The population configurations specify the size of each population according
to the information from the input species tree and the species name
corresponding to each population. Species names are stored as metadata in
each <a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> instance, with the metadata tag
“species_name”. Sampling configurations and growth rates are not specified
in the population configurations. The list of population configurations is
ordered according to the order of the corresponding extant species in a
<a class="reference external" href="https://en.wikipedia.org/wiki/Tree_traversal#Post-order_(LRN)">post-order tree traversal</a>.
The list of mass migration events and population parameter changes is
ordered by the time of the events, from young to old events.</p>
</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>, <a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)">list</a>)</p>
</dd>
<dt class="field-even">Warning</dt>
<dd class="field-even"><p>This function does not modify migration matrices. When the population
configurations and mass migration events returned by this function are used
to simulate with the <a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a> function, it should be ensured that
migration rates to source populations of mass migration events are zero
after the mass migration (viewed backwards in time).</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="populations">
<h2>Populations<a class="headerlink" href="#populations" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="msprime.Population">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">Population</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">initial_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">description</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">extra_metadata</span><span class="o">=</span><span class="default_value">NOTHING</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Population"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Population" title="Permalink to this definition">¶</a></dt>
<dd><p>Define a single population in a simulation.</p>
<dl class="field-list simple">
<dt class="field-odd">Variables</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>initial_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The absolute size of the population at time
zero. If not specified, or None, when the <a class="reference internal" href="#msprime.Demography" title="msprime.Demography"><code class="xref py py-class docutils literal notranslate"><span class="pre">Demography</span></code></a>
object is passed to the <code class="docutils literal notranslate"><span class="pre">demography</span></code> argument to <a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>
the default effective population size <code class="docutils literal notranslate"><span class="pre">Ne</span></code> will be used.</p></li>
<li><p><strong>growth_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The exponential growth rate of the
population per generation (forwards in time).
Growth rates can be negative. This is zero for a
constant population size, and positive for a population that has been
growing. Defaults to 0.</p></li>
<li><p><strong>name</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – The name of the population. If specified this must be a uniquely
identifying string.</p></li>
<li><p><strong>description</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#str" title="(in Python v3.9)"><em>str</em></a>) – a short description of the population</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="msprime.Population.from_old_style">
<em class="property">static </em><code class="sig-name descname">from_old_style</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">pop_config</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#Population.from_old_style"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.Population.from_old_style" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a Population object derived from the specified old-style
PopulationConfiguration.</p>
</dd></dl>

</dd></dl>

<div class="section" id="deprecated-api">
<h3>Deprecated API<a class="headerlink" href="#deprecated-api" title="Permalink to this headline">¶</a></h3>
<div class="admonition-todo admonition" id="id5">
<p class="admonition-title">Todo</p>
<p>Some introduction here to the deprecated 0.X API.</p>
</div>
<div class="admonition-todo admonition" id="id6">
<p class="admonition-title">Todo</p>
<p>refactor this text here which is drawn from the old api.rst
page.</p>
</div>
<p>Population structure is modelled in <code class="docutils literal notranslate"><span class="pre">msprime</span></code> by specifying a fixed number of
subpopulations, with the migration rates between those subpopulations defined by a migration
matrix. Each subpopulation has an <code class="docutils literal notranslate"><span class="pre">initial_size</span></code> that defines its absolute diploid size at
time zero and a per-generation <code class="docutils literal notranslate"><span class="pre">growth_rate</span></code> which specifies the exponential
growth rate of the sub-population. We must also define the number of genomes to
sample from each subpopulation. The number of populations and their initial
configuration is defined using the <code class="docutils literal notranslate"><span class="pre">population_configurations</span></code> parameter to
<a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>, which takes a list of <a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a>
instances. Population IDs are zero indexed, and correspond to their position in
the list.</p>
<p>Samples are drawn sequentially from populations in increasing order of
population ID. For example, if we specified an overall sample size of 6, and
specify that 2 samples are drawn from population 0 and 4 from population 1,
then samples 0 and 1 will be initially located in population 0, and
samples 2, 3, 4, and 5 will be drawn from population 2.</p>
<dl class="py class">
<dt id="msprime.PopulationConfiguration">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">PopulationConfiguration</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sample_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">initial_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">0.0</span></em>, <em class="sig-param"><span class="n">metadata</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#PopulationConfiguration"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.PopulationConfiguration" title="Permalink to this definition">¶</a></dt>
<dd><p>The initial configuration of a population (or deme) in a simulation.</p>
<div class="admonition-todo admonition" id="id7">
<p class="admonition-title">Todo</p>
<p>This is deprecated. Document as such.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>sample_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The number of initial samples that are drawn
from this population.</p></li>
<li><p><strong>initial_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The absolute size of the population at time
zero. Defaults to the reference population size <span class="math notranslate nohighlight">\(N_e\)</span>.</p></li>
<li><p><strong>growth_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The forwards-time exponential growth rate of the
population per generation. Growth rates can be negative. This is zero for a
constant population size, and positive for a population that has been
growing. Defaults to 0.</p></li>
<li><p><strong>metadata</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#dict" title="(in Python v3.9)"><em>dict</em></a>) – A JSON-encodable dictionary of metadata to associate
with the corresponding Population in the output tree sequence.
If not specified or None, no metadata is stored (i.e., an empty bytes array).
Note that this metadata is ignored when using the <code class="docutils literal notranslate"><span class="pre">from_ts</span></code> argument to
<a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>, as the population definitions in the tree sequence that
is used as the starting point take precedence.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="section" id="migration-rates">
<span id="sec-demography-migration"></span><h2>Migration rates<a class="headerlink" href="#migration-rates" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="demographic-events">
<span id="sec-demography-events"></span><h2>Demographic Events<a class="headerlink" href="#demographic-events" title="Permalink to this headline">¶</a></h2>
<p>Demographic events change some aspect of the population configuration
at some time in the past, and are specified using the <code class="docutils literal notranslate"><span class="pre">demographic_events</span></code>
parameter to <a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>. Each element of this list must be an
instance of one of the following demographic events
that are currently supported. Note that all times are measured in
generations, all sizes are absolute (i.e., <em>not</em> relative to <span class="math notranslate nohighlight">\(N_e\)</span>),
and all rates are per-generation.</p>
<p>Model change events are also considered to be demographic events;
see the <a class="reference internal" href="api.html#sec-api-simulation-models-change-events"><span class="std std-ref">Simulation model changes</span></a> section for details.</p>
<dl class="py class">
<dt id="msprime.PopulationParametersChange">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">PopulationParametersChange</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span></em>, <em class="sig-param"><span class="n">initial_size</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">growth_rate</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">population_id</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#PopulationParametersChange"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.PopulationParametersChange" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the demographic parameters of a population at a given time.</p>
<p>This event generalises the <code class="docutils literal notranslate"><span class="pre">-eg</span></code>, <code class="docutils literal notranslate"><span class="pre">-eG</span></code>, <code class="docutils literal notranslate"><span class="pre">-en</span></code> and <code class="docutils literal notranslate"><span class="pre">-eN</span></code>
options from <code class="docutils literal notranslate"><span class="pre">ms</span></code>. Note that unlike <code class="docutils literal notranslate"><span class="pre">ms</span></code> we do not automatically
set growth rates to zero when the population size is changed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The length of time ago at which this event
occurred.</p></li>
<li><p><strong>initial_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The absolute diploid size of the population
at the beginning of the time slice starting at <code class="docutils literal notranslate"><span class="pre">time</span></code>. If None,
this is calculated according to the initial population size and
growth rate over the preceding time slice.</p></li>
<li><p><strong>growth_rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The new per-generation growth rate. If None,
the growth rate is not changed. Defaults to None.</p></li>
<li><p><strong>population</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the population affected. If
<code class="docutils literal notranslate"><span class="pre">population</span></code> is None, the changes affect all populations
simultaneously.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.MigrationRateChange">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">MigrationRateChange</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span></em>, <em class="sig-param"><span class="n">rate</span></em>, <em class="sig-param"><span class="n">source</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">dest</span><span class="o">=</span><span class="default_value">- 1</span></em>, <em class="sig-param"><span class="n">matrix_index</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#MigrationRateChange"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.MigrationRateChange" title="Permalink to this definition">¶</a></dt>
<dd><p>Changes the rate of migration from one deme to another to a new value at a
specific time. Migration rates are specified in terms of the rate at which
lineages move from population <code class="docutils literal notranslate"><span class="pre">source</span></code> to <code class="docutils literal notranslate"><span class="pre">dest</span></code> during the progress of
the simulation. Note that <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> are from the perspective
of the coalescent process; please see the <a class="reference internal" href="api.html#sec-api-simulation-model"><span class="std std-ref">Simulation model</span></a>
section for more details on the interpretation of this migration model.</p>
<p>By default, <code class="docutils literal notranslate"><span class="pre">source=-1</span></code> and <code class="docutils literal notranslate"><span class="pre">dest=-1</span></code>, which results in all
non-diagonal elements of the migration matrix being changed to the new
rate. If <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> are specified, they must refer to valid
population IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The time at which this event occurs in generations.</p></li>
<li><p><strong>rate</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The new per-generation migration rate.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the source population.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the destination population.</p></li>
<li><p><strong>source</strong> – The source population ID.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.MassMigration">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">MassMigration</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span></em>, <em class="sig-param"><span class="n">source</span></em>, <em class="sig-param"><span class="n">dest</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">proportion</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">destination</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#MassMigration"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.MassMigration" title="Permalink to this definition">¶</a></dt>
<dd><p>A mass migration event in which some fraction of the population in one deme
(the <code class="docutils literal notranslate"><span class="pre">source</span></code>) simultaneously move to another deme (<code class="docutils literal notranslate"><span class="pre">dest</span></code>) during the
progress of the simulation. Each lineage currently present in the source
population moves to the destination population with probability equal to
<code class="docutils literal notranslate"><span class="pre">proportion</span></code>. Note that <code class="docutils literal notranslate"><span class="pre">source</span></code> and <code class="docutils literal notranslate"><span class="pre">dest</span></code> are from the perspective
of the coalescent process; please see the <a class="reference internal" href="api.html#sec-api-simulation-model"><span class="std std-ref">Simulation model</span></a>
section for more details on the interpretation of this migration model.</p>
<p>This event class generalises the population split (<code class="docutils literal notranslate"><span class="pre">-ej</span></code>) and
admixture (<code class="docutils literal notranslate"><span class="pre">-es</span></code>) events from <code class="docutils literal notranslate"><span class="pre">ms</span></code>. Note that MassMigrations
do <em>not</em> have any side effects on the migration matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The time at which this event occurs in generations.</p></li>
<li><p><strong>source</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the source population.</p></li>
<li><p><strong>dest</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The ID of the destination population.</p></li>
<li><p><strong>proportion</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The probability that any given lineage within
the source population migrates to the destination population.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class">
<dt id="msprime.CensusEvent">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">CensusEvent</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">time</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#CensusEvent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.CensusEvent" title="Permalink to this definition">¶</a></dt>
<dd><p>An event that adds a node to each branch of every tree at a given time
during the simulation. This may be used to record all ancestral haplotypes
present at that time, and to extract other information related to these
haplotypes: for instance to trace the local ancestry of a sample back to a
set of contemporaneous ancestors, or to assess whether a subset of samples
has coalesced more recently than the census time.
See the <a class="reference internal" href="tutorial.html#sec-tutorial-demography-census"><span class="std std-ref">tutorial</span></a> for an example.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>time</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The time at which this event occurs in generations.</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="debugging-demographic-models">
<h2>Debugging demographic models<a class="headerlink" href="#debugging-demographic-models" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="msprime.DemographyDebugger">
<em class="property">class </em><code class="sig-prename descclassname">msprime.</code><code class="sig-name descname">DemographyDebugger</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">demography</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">Ne</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">population_configurations</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">migration_matrix</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">demographic_events</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger" title="Permalink to this definition">¶</a></dt>
<dd><p>A class to facilitate debugging of population parameters and migration
rates in the past.</p>
<dl class="py method">
<dt id="msprime.DemographyDebugger.coalescence_rate_trajectory">
<code class="sig-name descname">coalescence_rate_trajectory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">steps</span></em>, <em class="sig-param"><span class="n">num_samples</span></em>, <em class="sig-param"><span class="n">min_pop_size</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">double_step_validation</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger.coalescence_rate_trajectory"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>This function will calculate the mean coalescence rates and proportions
of uncoalesced lineages between the lineages of the sample
configuration provided in <cite>num_samples</cite>, at each of the times ago
listed by steps, in this demographic model. The coalescence rate at
time t in the past is the average rate of coalescence of
as-yet-uncoalesed lineages, computed as follows: let <span class="math notranslate nohighlight">\(p(t)\)</span> be
the probability that the lineages of a randomly chosen pair of samples
has not yet coalesced by time <span class="math notranslate nohighlight">\(t\)</span>, let <span class="math notranslate nohighlight">\(p(z,t)\)</span> be the
probability that the lineages of a randomly chosen pair of samples has
not yet coalesced by time <span class="math notranslate nohighlight">\(t\)</span> <em>and</em> are both in population
<span class="math notranslate nohighlight">\(z\)</span>, and let <span class="math notranslate nohighlight">\(N(z,t)\)</span> be the diploid effective population
size of population <span class="math notranslate nohighlight">\(z\)</span> at time <span class="math notranslate nohighlight">\(t\)</span>. Then the mean
coalescence rate at time <span class="math notranslate nohighlight">\(t\)</span> is <span class="math notranslate nohighlight">\(r(t) = (\sum_z p(z,t) /
(2 * N(z,t)) / p(t)\)</span>.</p>
<p>The computation is done by approximating population size trajectories
with piecewise constant trajectories between each of the steps. For
this to be accurate, the distance between the steps must be small
enough so that (a) short epochs (e.g., bottlenecks) are not missed, and
(b) populations do not change in size too much over that time, if they
are growing or shrinking. This function optionally provides a simple
check of this approximation by recomputing the coalescence rates on a
grid of steps twice as fine and throwing a warning if the resulting
values do not match to a relative tolerance of 0.001.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The times ago at which coalescence rates will be computed.</p></li>
<li><p><strong>num_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of the same length as the number
of populations, so that <cite>num_samples[j]</cite> is the number of sampled
chromosomes in subpopulation <cite>j</cite>.</p></li>
<li><p><strong>min_pop_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The smallest allowed population size during
computation of coalescent rates (i.e., coalescence rates are actually
1 / (2 * max(min_pop_size, N(z,t))). Spurious very small population sizes
can occur in models where populations grow exponentially but are unused
before some time in the past, and lead to floating point error.
This should be set to a value smaller than the smallest
desired population size in the model.</p></li>
<li><p><strong>double_step_validation</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)"><em>bool</em></a>) – Whether to perform the check that
step sizes are sufficiently small, as described above. This is highly
recommended, and will take at most four times the computation.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A tuple of arrays whose jth elements, respectively, are the
coalescence rate at the jth time point (denoted r(t[j]) above),
and the probablility that a randomly chosen pair of lineages has
not yet coalesced (denoted p(t[j]) above).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>(numpy.array, numpy.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.epoch_times">
<em class="property">property </em><code class="sig-name descname">epoch_times</code><a class="headerlink" href="#msprime.DemographyDebugger.epoch_times" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns array of epoch times defined by the demographic model</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.lineage_probabilities">
<code class="sig-name descname">lineage_probabilities</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">steps</span></em>, <em class="sig-param"><span class="n">sample_time</span><span class="o">=</span><span class="default_value">0</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger.lineage_probabilities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger.lineage_probabilities" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns an array such that P[j, a, b] is the probability that a lineage that
started in population a at time sample_time is in population b at time steps[j]
ago.</p>
<p>This function reports sampling probabilities _before_ mass migration events
at a step time, if a mass migration event occurs at one of those times.
Migrations will then effect the next time step.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of times to compute probabilities.</p></li>
<li><p><strong>sample_time</strong> – The time of sampling of the lineage. For any times in steps
that are more recent than sample_time, the probability of finding the
lineage in any population is zero.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>An array of dimension len(steps) by num pops by num_pops.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.mean_coalescence_time">
<code class="sig-name descname">mean_coalescence_time</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">num_samples</span></em>, <em class="sig-param"><span class="n">min_pop_size</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">steps</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">rtol</span><span class="o">=</span><span class="default_value">0.005</span></em>, <em class="sig-param"><span class="n">max_iter</span><span class="o">=</span><span class="default_value">12</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger.mean_coalescence_time"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger.mean_coalescence_time" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the mean time until coalescence between lineages of two samples drawn
from the sample configuration specified in <cite>num_samples</cite>. This is done using
<a class="reference internal" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">coalescence_rate_trajectory</span></code></a>
to compute the probability that the lineages have not yet coalesced by time <cite>t</cite>,
and using these to approximate <span class="math notranslate nohighlight">\(E[T] = \int_t^\infty P(T &gt; t) dt\)</span>,
where <span class="math notranslate nohighlight">\(T\)</span> is the coalescence time. See
<a class="reference internal" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">coalescence_rate_trajectory</span></code></a>
for more details.</p>
<p>To compute this, an adequate time discretization must be arrived at
by iteratively extending or refining the current discretization.
Debugging information about numerical convergence of this procedure is
logged using the Python <a class="reference external" href="https://docs.python.org/3/library/logging.html#module-logging" title="(in Python v3.9)"><code class="xref py py-mod docutils literal notranslate"><span class="pre">logging</span></code></a> infrastructure. To make it appear, using
the <code class="xref py py-mod docutils literal notranslate"><span class="pre">daiquiri</span></code> module, do for instance:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">daiquiri</span>

<span class="n">daiquiri</span><span class="o">.</span><span class="n">setup</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="s2">&quot;DEBUG&quot;</span><span class="p">)</span>
<span class="n">debugger</span><span class="o">.</span><span class="n">mean_coalescence_time</span><span class="p">([</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
<p>will print this debugging information to stderr. Briefly, this outputs
iteration number, mean coalescence time, maximum difference in probabilty
of not having coalesced yet, difference to last coalescence time,
probability of not having coalesced by the final time point, and
whether the last iteration was an extension or refinement.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>num_samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of the same length as the number
of populations, so that <cite>num_samples[j]</cite> is the number of sampled
chromosomes in subpopulation <cite>j</cite>.</p></li>
<li><p><strong>min_pop_size</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – See <a class="reference internal" href="#msprime.DemographyDebugger.coalescence_rate_trajectory" title="msprime.DemographyDebugger.coalescence_rate_trajectory"><code class="xref py py-meth docutils literal notranslate"><span class="pre">coalescence_rate_trajectory</span></code></a>.</p></li>
<li><p><strong>steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – The time discretization to start out with (by default,
picks something based on epoch times).</p></li>
<li><p><strong>rtol</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)"><em>float</em></a>) – The relative tolerance to determine mean coalescence time
to (used to decide when to stop subdividing the steps).</p></li>
<li><p><strong>max_iter</strong> (<a class="reference external" href="https://docs.python.org/3/library/functions.html#int" title="(in Python v3.9)"><em>int</em></a>) – The maximum number of times to subdivide the steps.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The mean coalescence time (a number).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference external" href="https://docs.python.org/3/library/functions.html#float" title="(in Python v3.9)">float</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.num_epochs">
<em class="property">property </em><code class="sig-name descname">num_epochs</code><a class="headerlink" href="#msprime.DemographyDebugger.num_epochs" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the number of epochs defined by the demographic model.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.population_size_history">
<em class="property">property </em><code class="sig-name descname">population_size_history</code><a class="headerlink" href="#msprime.DemographyDebugger.population_size_history" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a (num_pops, num_epochs) numpy array giving the starting population size
for each population in each epoch.</p>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.population_size_trajectory">
<code class="sig-name descname">population_size_trajectory</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">steps</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger.population_size_trajectory"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger.population_size_trajectory" title="Permalink to this definition">¶</a></dt>
<dd><p>This function returns an array of per-population effective population sizes,
as defined by the demographic model. These are the <cite>initial_size</cite>
parameters of the model, modified by any population growth rates.
The sizes are computed at the time points given by <cite>steps</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>steps</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – List of times ago at which the population
size will be computed.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns a numpy array of population sizes, with one column per
population, whose [i,j]th entry is the size of population
j at time steps[i] ago.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.possible_lineage_locations">
<code class="sig-name descname">possible_lineage_locations</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">samples</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger.possible_lineage_locations"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger.possible_lineage_locations" title="Permalink to this definition">¶</a></dt>
<dd><p>Given the sampling configuration, this function determines when lineages are
possibly found within each population over epochs defined by demographic events
and sampling times. If no sampling configuration is given, we assume we sample
lineages from every population at time zero. The samples are specified by a list
of msprime Sample objects, so that possible ancient samples may be accounted for.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>samples</strong> (<a class="reference external" href="https://docs.python.org/3/library/stdtypes.html#list" title="(in Python v3.9)"><em>list</em></a>) – A list of msprime Sample objects, which specify their
populations and times.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Returns a dictionary with epoch intervals as keys whose values are a
list with length equal to the number of populations with True and False
indicating which populations could possibly contain lineages over that
epoch. The epoch intervals are given by tuples: (epoch start, epoch end).
The first epoch necessarily starts at time 0, and the final epoch has end
time of infinity.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="msprime.DemographyDebugger.print_history">
<code class="sig-name descname">print_history</code><span class="sig-paren">(</span><em class="sig-param">output=&lt;_io.TextIOWrapper name='&lt;stdout&gt;' mode='w' encoding='utf-8'&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/msprime/demography.html#DemographyDebugger.print_history"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#msprime.DemographyDebugger.print_history" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints a summary of the history of the populations.</p>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="species-trees">
<h2>Species trees<a class="headerlink" href="#species-trees" title="Permalink to this headline">¶</a></h2>
<div class="admonition-todo admonition" id="id8">
<p class="admonition-title">Todo</p>
<p>Review this documentation which is a combination of the old API and
tutorial docs.</p>
</div>
<p>Species trees hold information about the sequence and the times at which species
diverged from each other. Viewed backwards in time, divergence events are equivalent
to mass migration events in which all lineages from one population move to another
population. The history of a set of populations can thus be modelled according to
a given species tree. To faciliate the specification of the model,
<a class="reference internal" href="#msprime.parse_species_tree" title="msprime.parse_species_tree"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_species_tree()</span></code></a> parses a species tree and returns the mass migration
events corresponding to all species divergence events in the tree, together with
population configurations that specify population sizes and names.</p>
<p>When species trees are estimated with a program like <a class="reference external" href="https://academic.oup.com/mbe/article/34/8/2101/3738283">StarBEAST</a> they can further
contain estimates on the population sizes of extant and ancestral species.
<a class="reference internal" href="#msprime.parse_starbeast" title="msprime.parse_starbeast"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_starbeast()</span></code></a> parses species trees estimated with StarBEAST and uses
these estimates to define the population configurations.</p>
<p>Note that when the species tree has branch lengths not in units of generations but
in units of years or millions of years (which is common), a generation time in years
is required for parsing.</p>
<div class="admonition-todo admonition" id="id10">
<p class="admonition-title">Todo</p>
<p>refactor this content to use the new conventions and jupyter.</p>
</div>
<p>Models used in <code class="docutils literal notranslate"><span class="pre">msprime</span></code> for simulation can be designed to approximate the
diversification history of a group of diverging species, by defining, for each species
divergence, a mass migration event at which all lineages from one population move
into another population. To faciliate the specification of these mass migration events
it is possible to parse a species tree and generate the set of mass migration events
automatically.</p>
<p>To be parseable, a species tree must be encoded in
<a class="reference external" href="https://en.wikipedia.org/wiki/Newick_format">Newick</a> format, with named leaves and
branch lengths. One example of a parseable species tree in Newick format is
<code class="docutils literal notranslate"><span class="pre">(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)</span></code>. When visualized
in a software like <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree/">FigTree</a>, this tree
appears as shown below:</p>
<a class="reference internal image-reference" href="_images/primates.svg"><img alt="A species tree with branch lengths." src="_images/primates.svg" width="400px" /></a>
<p>In the above figure, numbers written on branches indicate the lengths of these branches.
In this case, the units of the branch lengths are millions of years, which is common
for species trees; however, trees with branch lengths in units of years or generations
can also be parsed. When the branch lengths are in units of millions of years or in units
of years (i.e., not in units of generations), a generation time in number of years must
be provided so that the simulation model can be set up. In addition, a population size
is required. With the species tree, a generation time, and the population size, the model
can be generated:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">msprime</span>
<span class="n">pop_configs</span><span class="p">,</span> <span class="n">demographic_events</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">parse_species_tree</span><span class="p">(</span>
        <span class="n">tree</span><span class="o">=</span><span class="s2">&quot;(((human:5.6,chimpanzee:5.6):3.0,gorilla:8.6):9.4,orangutan:18.0)&quot;</span><span class="p">,</span>
        <span class="n">Ne</span><span class="o">=</span><span class="mi">10000</span><span class="p">,</span>
        <span class="n">branch_length_units</span><span class="o">=</span><span class="s2">&quot;myr&quot;</span><span class="p">,</span>
        <span class="n">generation_time</span><span class="o">=</span><span class="mi">28</span><span class="p">)</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">parse_species_tree()</span></code> method returns a tuple of two lists:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">pop_configs</span><span class="p">))</span>
<span class="c1"># &lt;class &#39;list&#39;&gt;</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">demographic_events</span><span class="p">))</span>
<span class="c1"># &lt;class &#39;list&#39;&gt;</span>
</pre></div>
</div>
<p>The first of these two lists contains instances of <a class="reference internal" href="#msprime.PopulationConfiguration" title="msprime.PopulationConfiguration"><code class="xref py py-class docutils literal notranslate"><span class="pre">PopulationConfiguration</span></code></a> and
the second contains instances of <a class="reference internal" href="#msprime.MassMigration" title="msprime.MassMigration"><code class="xref py py-class docutils literal notranslate"><span class="pre">MassMigration</span></code></a>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">pop_configs</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="c1"># &lt;class &#39;msprime.simulations.PopulationConfiguration&#39;&gt;</span>
<span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">demographic_events</span><span class="p">[</span><span class="mi">0</span><span class="p">]))</span>
<span class="c1"># &lt;class &#39;msprime.simulations.MassMigration&#39;&gt;</span>
</pre></div>
</div>
<p>The mass migration events are ordered by the time of the event and they thus specify
the sequence in which lineages are moved from source to destination populations:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">demographic_events</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
<span class="c1"># Mass migration: Lineages moved with probability 1.0 backwards in time with source 1 &amp; dest 0</span>
                     <span class="p">(</span><span class="n">equivalent</span> <span class="n">to</span> <span class="n">migration</span> <span class="kn">from</span> <span class="mi">0</span> <span class="n">to</span> <span class="mi">1</span> <span class="n">forwards</span> <span class="ow">in</span> <span class="n">time</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">demographic_events</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
<span class="c1"># Mass migration: Lineages moved with probability 1.0 backwards in time with source 2 &amp; dest 0</span>
                     <span class="p">(</span><span class="n">equivalent</span> <span class="n">to</span> <span class="n">migration</span> <span class="kn">from</span> <span class="mi">0</span> <span class="n">to</span> <span class="mi">2</span> <span class="n">forwards</span> <span class="ow">in</span> <span class="n">time</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">demographic_events</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
<span class="c1"># Mass migration: Lineages moved with probability 1.0 backwards in time with source 3 &amp; dest 0</span>
                     <span class="p">(</span><span class="n">equivalent</span> <span class="n">to</span> <span class="n">migration</span> <span class="kn">from</span> <span class="mi">0</span> <span class="n">to</span> <span class="mi">3</span> <span class="n">forwards</span> <span class="ow">in</span> <span class="n">time</span><span class="p">)</span>
</pre></div>
</div>
<p>The above output indicates that — viewed backwards in time — lineages from
populations 1, 2, and 3 are consecutively moved into population 0. Viewed forwards in
time instead, this means that population 3 is the first to diverge, followed by
population 2 and finally the divergence between populations 0 and 1. This
sequence of divergences corresponds to the species tree if population 3 is orangutan
and populations 2, 1, and 0 are gorilla, chimpanzee, and human, respectively. While
the parsed species names are not used as population labels, they are included in the
population configurations in the form of metadata, with the “species_name” tag:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">([</span><span class="n">pop_config</span><span class="o">.</span><span class="n">metadata</span> <span class="k">for</span> <span class="n">pop_config</span> <span class="ow">in</span> <span class="n">pop_configs</span><span class="p">])</span>
<span class="c1"># [{&#39;species_name&#39;: &#39;human&#39;}, {&#39;species_name&#39;: &#39;chimpanzee&#39;}, {&#39;species_name&#39;: &#39;gorilla&#39;}, {&#39;species_name&#39;: &#39;orangutan&#39;}]</span>
</pre></div>
</div>
<p>As the above output shows, the information on the topology of the species tree is
fully included in the set of population configurations and mass migration events.
It also illustrates that it is always the left one (when viewed from the root
towards the tips of the species tree in the tree figure above) of the two populations
descending from a divergence event that is used as the destination population in
mass migration events.</p>
<p>The population configurations also define the population size:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">([</span><span class="n">pop_config</span><span class="o">.</span><span class="n">initial_size</span> <span class="k">for</span> <span class="n">pop_config</span> <span class="ow">in</span> <span class="n">pop_configs</span><span class="p">])</span>
<span class="c1"># [10000.0, 10000.0, 10000.0, 10000.0]</span>
</pre></div>
</div>
<p>The population size is 10,000 because this value was specified for Ne when
calling <a class="reference internal" href="#msprime.parse_species_tree" title="msprime.parse_species_tree"><code class="xref py py-func docutils literal notranslate"><span class="pre">parse_species_tree()</span></code></a>. The growth rates are zero for all populations,
meaning that they all have constant population sizes:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">([</span><span class="n">pop_config</span><span class="o">.</span><span class="n">growth_rate</span> <span class="k">for</span> <span class="n">pop_config</span> <span class="ow">in</span> <span class="n">pop_configs</span><span class="p">])</span>
<span class="c1"># [0.0, 0.0, 0.0, 0.0]</span>
</pre></div>
</div>
<p>To simulate under the model corresponding to the species tree, the population
configurations and mass migration events are used as input for
<a class="reference internal" href="ancestry.html#msprime.simulate" title="msprime.simulate"><code class="xref py py-func docutils literal notranslate"><span class="pre">simulate()</span></code></a>. We can specify the genomes to sample either by using the
<code class="docutils literal notranslate"><span class="pre">samples</span></code> parameter or by setting a sample size for each population:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pop_config</span> <span class="ow">in</span> <span class="n">pop_configs</span><span class="p">:</span>
    <span class="n">pop_config</span><span class="o">.</span><span class="n">sample_size</span> <span class="o">=</span> <span class="mi">2</span>
<span class="c1"># pop_configs now has 2 genomes sampled from each population at time 0. Equivalent to</span>
<span class="c1"># msprime.simulate(samples=[(0,0), (0,0), (1,0), (1,0), (2,0), (2,0), (3,0), (3,0)], ...)</span>
<span class="n">tree_sequence</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">simulate</span><span class="p">(</span>
        <span class="n">population_configurations</span><span class="o">=</span><span class="n">pop_configs</span><span class="p">,</span>
        <span class="n">demographic_events</span><span class="o">=</span><span class="n">demographic_events</span><span class="p">)</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">tree_sequence</span><span class="o">.</span><span class="n">first</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">draw</span><span class="p">(</span><span class="nb">format</span><span class="o">=</span><span class="s2">&quot;unicode&quot;</span><span class="p">))</span>
<span class="c1">#    14</span>
<span class="c1">#  ┏━━┻━━━┓</span>
<span class="c1">#  ┃     13</span>
<span class="c1">#  ┃   ┏━━┻━━┓</span>
<span class="c1">#  ┃   ┃    12</span>
<span class="c1">#  ┃   ┃   ┏━┻━┓</span>
<span class="c1">#  ┃   ┃   ┃  11</span>
<span class="c1">#  ┃   ┃   ┃  ┏┻┓</span>
<span class="c1"># 10   ┃   ┃  ┃ ┃</span>
<span class="c1"># ┏┻┓  ┃   ┃  ┃ ┃</span>
<span class="c1"># ┃ ┃  ┃   9  ┃ ┃</span>
<span class="c1"># ┃ ┃  ┃  ┏┻┓ ┃ ┃</span>
<span class="c1"># ┃ ┃  8  ┃ ┃ ┃ ┃</span>
<span class="c1"># ┃ ┃ ┏┻┓ ┃ ┃ ┃ ┃</span>
<span class="c1"># 6 7 4 5 2 3 0 1</span>
</pre></div>
</div>
<p>The correspondence between the model and the species tree can also be verified
by using the demography debugger:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dd</span> <span class="o">=</span> <span class="n">msprime</span><span class="o">.</span><span class="n">DemographyDebugger</span><span class="p">(</span>
    <span class="n">population_configurations</span><span class="o">=</span><span class="n">pop_configs</span><span class="p">,</span>
    <span class="n">demographic_events</span><span class="o">=</span><span class="n">demographic_events</span><span class="p">)</span>
<span class="n">dd</span><span class="o">.</span><span class="n">print_history</span><span class="p">()</span>
<span class="c1"># Model =  hudson(reference_size=1)</span>
<span class="c1"># ================================</span>
<span class="c1"># Epoch: 0 -- 200000.0 generations</span>
<span class="c1"># ================================</span>
<span class="c1">#      start     end      growth_rate |     0        1        2        3</span>
<span class="c1">#    -------- --------       -------- | -------- -------- -------- --------</span>
<span class="c1"># 0 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 1 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 2 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 3 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1">#</span>
<span class="c1"># Events @ generation 200000.0</span>
<span class="c1">#    - Mass migration: Lineages moved with probability 1.0 backwards in time with source 1 &amp; dest 0</span>
<span class="c1">#                      (equivalent to migration from 0 to 1 forwards in time)</span>
<span class="c1"># =================================================</span>
<span class="c1"># Epoch: 200000.0 -- 307142.85714285716 generations</span>
<span class="c1"># =================================================</span>
<span class="c1">#      start     end      growth_rate |     0        1        2        3</span>
<span class="c1">#    -------- --------       -------- | -------- -------- -------- --------</span>
<span class="c1"># 0 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 1 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 2 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 3 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1">#</span>
<span class="c1"># Events @ generation 307142.85714285716</span>
<span class="c1">#    - Mass migration: Lineages moved with probability 1.0 backwards in time with source 2 &amp; dest 0</span>
<span class="c1">#                      (equivalent to migration from 0 to 2 forwards in time)</span>
<span class="c1"># =========================================================</span>
<span class="c1"># Epoch: 307142.85714285716 -- 642857.142857143 generations</span>
<span class="c1"># =========================================================</span>
<span class="c1">#      start     end      growth_rate |     0        1        2        3</span>
<span class="c1">#    -------- --------       -------- | -------- -------- -------- --------</span>
<span class="c1"># 0 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 1 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 2 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 3 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1">#</span>
<span class="c1"># Events @ generation 642857.142857143</span>
<span class="c1">#    - Mass migration: Lineages moved with probability 1.0 backwards in time with source 3 &amp; dest 0</span>
<span class="c1">#                      (equivalent to migration from 0 to 3 forwards in time)</span>
<span class="c1"># ==========================================</span>
<span class="c1"># Epoch: 642857.142857143 -- inf generations</span>
<span class="c1"># ==========================================</span>
<span class="c1">#      start     end      growth_rate |     0        1        2        3</span>
<span class="c1">#    -------- --------       -------- | -------- -------- -------- --------</span>
<span class="c1"># 0 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 1 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 2 |  1e+04    1e+04               0 |     0        0        0        0</span>
<span class="c1"># 3 |  1e+04    1e+04               0 |     0        0        0        0</span>
</pre></div>
</div>
<p>The epoch boundaries 200000, 307142.9, and 642857.1 correspond to the species
divergence times 5.6, 8.6, and 18.0 after converting the branch length units
of the species tree from millions of years to generations with the specified
generation time of 28 years.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="likelihoods.html" class="btn btn-neutral float-right" title="Likelihoods" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="mutations.html" class="btn btn-neutral float-left" title="Mutations" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2015-2020, Tskit Developers

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>